src/Umbraco.Web.BackOffice/AspNetCore/UmbracoBackOfficeServiceCollectionExtensions.cs

```diff
@@ -30,53 +42,116 @@ public static IServiceCollection AddUmbracoConfiguration(this IServiceCollection
+            var globalSettings = configs.Global();
+            var connStrings = configs.ConnectionStrings();
+            var appSettingMainDomLock = globalSettings.MainDomLock;
+            var mainDomLock = appSettingMainDomLock == "SqlMainDomLock"
+                ? (IMainDomLock)new SqlMainDomLock(logger, globalSettings, connStrings, dbProviderFactoryCreator)
+                : new MainDomSemaphoreLock(logger, hostingEnvironment);
+
+            var mainDom = new MainDom(logger, hostingEnvironment, mainDomLock);
+
+            var coreRuntime = new CoreRuntime(configs, umbracoVersion, ioHelper, logger, profiler, new AspNetCoreBootPermissionsChecker(),
+                hostingEnvironment, backOfficeInfo, dbProviderFactoryCreator, mainDom, typeFinder);
 
-        public static IServiceCollection CreateCompositionRoot(
-            this IServiceCollection services,
-            IHttpContextAccessor httpContextAccessor,
-            IWebHostEnvironment webHostEnvironment,
-            IHostApplicationLifetime hostApplicationLifetime,
-            Configs configs)
+            return coreRuntime;
+        }
+
+        private static void CreateCompositionRoot(IServiceCollection services)
         {
+            // TODO: This isn't the best to have to resolve the services now but to avoid this will
+            // require quite a lot of re-work. 
+            var serviceProvider = services.BuildServiceProvider();
+
+            var httpContextAccessor = serviceProvider.GetRequiredService<IHttpContextAccessor>();
+            var webHostEnvironment = serviceProvider.GetRequiredService<IWebHostEnvironment>();
+            var hostApplicationLifetime = serviceProvider.GetRequiredService<IHostApplicationLifetime>();
+
+            var configs = serviceProvider.GetService<Configs>();
+            if (configs == null)
+                throw new InvalidOperationException($"Could not resolve type {typeof(Configs)} from the container, ensure {nameof(AddUmbracoConfiguration)} is called before calling {nameof(AddUmbracoCore)}");
+
             var hostingSettings = configs.Hosting();
             var coreDebug = configs.CoreDebug();
             var globalSettings = configs.Global();
 
-            var hostingEnvironment = new AspNetCoreHostingEnvironment(hostingSettings, webHostEnvironment,
-                httpContextAccessor, hostApplicationLifetime);
+            var hostingEnvironment = new AspNetCoreHostingEnvironment(hostingSettings, webHostEnvironment, httpContextAccessor, hostApplicationLifetime);
             var ioHelper = new IOHelper(hostingEnvironment, globalSettings);
             var logger = SerilogLogger.CreateWithDefaultConfiguration(hostingEnvironment,
                 new AspNetCoreSessionIdResolver(httpContextAccessor),
-                () => services.BuildServiceProvider().GetService<IRequestCache>(), coreDebug, ioHelper,
+                () => serviceProvider.GetService<IRequestCache>(), coreDebug, ioHelper,
```

Comment:
I don't think the `serviceProvider`, build at that time, has a `IRequestCache` availble, that why it was done the other way.

Reply:
ah right I see. I'll change back. I've added a couple more notes in a newer branch about the whole BuildServiceProvider() while we are building the container. I'm not convinced this will work correctly because it will end up being different instances of a service that is supposed to be a singleton. The main problem is going to be `IHostApplicationLifetime` I think. Each time you call `BuildServiceProvider` and then resolve a service from that, even if the service is registered as a singleton, it will be a new instance for each service provider created which is really not what we want to have happen. I don't think we can solve this in this PR but we'll need to considering this going forward.